File: Frameworks/flutter-local/test/models/ods_style_hint_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:ods_flutter_local/models/ods_style_hint.dart';

void main() {
  // =========================================================================
  // fromJson
  // =========================================================================

  group('OdsStyleHint.fromJson', () {
    test('fromJson with null returns empty hints', () {
      final h = OdsStyleHint.fromJson(null);
      expect(h.hints, isEmpty);
    });

    test('fromJson with valid map passes through', () {
      final h = OdsStyleHint.fromJson({'variant': 'heading', 'color': 'red'});
      expect(h.hints['variant'], 'heading');
      expect(h.hints['color'], 'red');
    });
  });

  // =========================================================================
  // get<T>
  // =========================================================================

  group('get<T>', () {
    test('extracts typed value', () {
      final h = OdsStyleHint({'count': 42, 'name': 'test'});
      expect(h.get<int>('count'), 42);
      expect(h.get<String>('name'), 'test');
    });

    test('returns null for missing key', () {
      final h = OdsStyleHint({'a': 1});
      expect(h.get<String>('missing'), isNull);
    });

    test('returns null for wrong type', () {
      final h = OdsStyleHint({'count': 42});
      expect(h.get<String>('count'), isNull);
    });
  });

  // =========================================================================
  // Convenience accessors
  // =========================================================================

  group('convenience accessors', () {
    test('variant returns string', () {
      final h = OdsStyleHint({'variant': 'heading'});
      expect(h.variant, 'heading');
    });

    test('emphasis returns string', () {
      final h = OdsStyleHint({'emphasis': 'primary'});
      expect(h.emphasis, 'primary');
    });

    test('align returns string', () {
      final h = OdsStyleHint({'align': 'center'});
      expect(h.align, 'center');
    });

    test('color returns string', () {
      final h = OdsStyleHint({'color': 'red'});
      expect(h.color, 'red');
    });

    test('icon returns string', () {
      final h = OdsStyleHint({'icon': 'star'});
      expect(h.icon, 'star');
    });

    test('size returns string', () {
      final h = OdsStyleHint({'size': 'compact'});
      expect(h.size, 'compact');
    });

    test('density returns string', () {
      final h = OdsStyleHint({'density': 'comfortable'});
      expect(h.density, 'comfortable');
    });

    test('accessors return null when key is missing', () {
      final h = OdsStyleHint({});
      expect(h.variant, isNull);
      expect(h.emphasis, isNull);
      expect(h.align, isNull);
      expect(h.color, isNull);
      expect(h.icon, isNull);
      expect(h.size, isNull);
      expect(h.density, isNull);
    });
  });

  // =========================================================================
  // elevation
  // =========================================================================

  group('elevation', () {
    test('returns int directly', () {
      final h = OdsStyleHint({'elevation': 2});
      expect(h.elevation, 2);
    });

    test('returns int from double (floor)', () {
      final h = OdsStyleHint({'elevation': 2.7});
      expect(h.elevation, 2);
    });

    test('returns null for non-number', () {
      final h = OdsStyleHint({'elevation': 'high'});
      expect(h.elevation, isNull);
    });

    test('returns null when missing', () {
      final h = OdsStyleHint({});
      expect(h.elevation, isNull);
    });
  });

  // =========================================================================
  // isEmpty
  // =========================================================================

  group('isEmpty', () {
    test('is true for empty hints', () {
      final h = OdsStyleHint({});
      expect(h.isEmpty, true);
    });

    test('is true for fromJson null', () {
      final h = OdsStyleHint.fromJson(null);
      expect(h.isEmpty, true);
    });

    test('is false for non-empty hints', () {
      final h = OdsStyleHint({'variant': 'heading'});
      expect(h.isEmpty, false);
    });
  });
}

File: Frameworks/react-web/tests/unit/models/ods-style-hint.test.ts
import { describe, it, expect } from 'vitest'
import {
  parseStyleHint,
  getHint,
  hintVariant,
  hintEmphasis,
  hintAlign,
  hintColor,
  hintIcon,
  hintSize,
  hintDensity,
  hintElevation,
} from '../../../src/models/ods-style-hint.ts'

// ===========================================================================
// OdsStyleHint model tests
// ===========================================================================

describe('parseStyleHint', () => {
  // -------------------------------------------------------------------------
  // Null / invalid input
  // -------------------------------------------------------------------------

  it('returns empty object for null', () => {
    expect(parseStyleHint(null)).toEqual({})
  })

  it('returns empty object for undefined', () => {
    expect(parseStyleHint(undefined)).toEqual({})
  })

  it('returns empty object for array', () => {
    expect(parseStyleHint([1, 2, 3])).toEqual({})
  })

  it('returns empty object for string', () => {
    expect(parseStyleHint('not-an-object')).toEqual({})
  })

  it('returns empty object for number', () => {
    expect(parseStyleHint(42)).toEqual({})
  })

  it('returns empty object for boolean', () => {
    expect(parseStyleHint(true)).toEqual({})
  })

  // -------------------------------------------------------------------------
  // Valid object input
  // -------------------------------------------------------------------------

  it('passes through a valid object', () => {
    const hint = { variant: 'outlined', color: 'primary' }
    const result = parseStyleHint(hint)
    expect(result).toEqual({ variant: 'outlined', color: 'primary' })
  })

  it('passes through an empty object', () => {
    expect(parseStyleHint({})).toEqual({})
  })
})

// ===========================================================================
// getHint
// ===========================================================================

describe('getHint', () => {
  it('extracts a typed value', () => {
    const hint = { variant: 'filled', size: 'lg' }
    expect(getHint<string>(hint, 'variant')).toBe('filled')
  })

  it('returns undefined for missing key', () => {
    const hint = { variant: 'filled' }
    expect(getHint<string>(hint, 'color')).toBeUndefined()
  })

  it('extracts a number value', () => {
    const hint = { elevation: 3 }
    expect(getHint<number>(hint, 'elevation')).toBe(3)
  })
})

// ===========================================================================
// Convenience accessors
// ===========================================================================

describe('convenience accessors', () => {
  const hint = {
    variant: 'outlined',
    emphasis: 'high',
    align: 'center',
    color: 'primary',
    icon: 'star',
    size: 'lg',
    density: 'compact',
    elevation: 2.7,
  }

  it('hintVariant returns variant', () => {
    expect(hintVariant(hint)).toBe('outlined')
  })

  it('hintEmphasis returns emphasis', () => {
    expect(hintEmphasis(hint)).toBe('high')
  })

  it('hintAlign returns align', () => {
    expect(hintAlign(hint)).toBe('center')
  })

  it('hintColor returns color', () => {
    expect(hintColor(hint)).toBe('primary')
  })

  it('hintIcon returns icon', () => {
    expect(hintIcon(hint)).toBe('star')
  })

  it('hintSize returns size', () => {
    expect(hintSize(hint)).toBe('lg')
  })

  it('hintDensity returns density', () => {
    expect(hintDensity(hint)).toBe('compact')
  })

  // -------------------------------------------------------------------------
  // hintElevation
  // -------------------------------------------------------------------------

  it('hintElevation returns floored number', () => {
    expect(hintElevation(hint)).toBe(2)
  })

  it('hintElevation returns integer unchanged', () => {
    expect(hintElevation({ elevation: 4 })).toBe(4)
  })

  it('hintElevation returns undefined for non-number', () => {
    expect(hintElevation({ elevation: 'high' })).toBeUndefined()
  })

  it('hintElevation returns undefined when missing', () => {
    expect(hintElevation({})).toBeUndefined()
  })

  // -------------------------------------------------------------------------
  // Accessors return undefined for missing keys
  // -------------------------------------------------------------------------

  it('hintVariant returns undefined when missing', () => {
    expect(hintVariant({})).toBeUndefined()
  })

  it('hintEmphasis returns undefined when missing', () => {
    expect(hintEmphasis({})).toBeUndefined()
  })

  it('hintAlign returns undefined when missing', () => {
    expect(hintAlign({})).toBeUndefined()
  })

  it('hintColor returns undefined when missing', () => {
    expect(hintColor({})).toBeUndefined()
  })

  it('hintIcon returns undefined when missing', () => {
    expect(hintIcon({})).toBeUndefined()
  })

  it('hintSize returns undefined when missing', () => {
    expect(hintSize({})).toBeUndefined()
  })

  it('hintDensity returns undefined when missing', () => {
    expect(hintDensity({})).toBeUndefined()
  })
})
